### ISSUE TITLE
glfwSwapInterval having no effect

### ISSUE THREAD

**Issue by HugoPeters1024** [glfw/glfw#1559]
## Problem
The call to glfwSwapInterval(1) does not produce the expected 60 fps experience on my gtx680m. Instead I get an enormous swap rate as one would expect using glfwSwapInterval(0). The problem only occurs when using my gtx680m (running on proprietary nvidia-driver-430 (recommended)). On my IntelHd everything works as expected

## Related information
I found older forum threads relating to this issue suggesting you needed to define a certain macro and then recompile glfw. Some people suggested however that this was already fixed since 3.1.2*
*https://stackoverflow.com/questions/16285546/glfwswapinterval1-fails-to-enable-vsync (2nd answer)

Versions
glfw 3.2.1 installed via apt
ubuntu 18.04

I would love to provide more information when needed but please keep in mind that I am a c++ beginner and have never compiled a library myself, so be gentle :)

**Comment by HugoPeters1024** [comment]
Problem was not reproduced on glfw 3.3.2 compiled from source, closing

**Comment by elmindreda** [comment]
Thank you for the follow-up!

**Comment by NicolayP** [comment]
Hi I get a similar issue with glfw 3.4.0 compiled from source

**Comment by HugoPeters1024** [comment]
I forgot about this, but I do still have this bug as well. I think I might have ran it on my onboard intel accidentally thinking it was resolved. As a work around I use a blocking while loop until 1/60th of a second has passed but that is of course not ideal nor very nice to the rest of the processes on the machine.

**Comment by NicolayP** [comment]
I'm doing the same in the mean time but as you said not a nice solution.

btw i'm on ubuntu 18.04 as well

**Comment by beaumanvienna** [comment]
Hello! Could I please join in? I have the same problem. The environment is Ubuntu Focal, X11,  and glfw is compiled from upstream.
I build glfw with `cmake ..`. I make sure to first create a context and only then set up the swap interval. For both commands, I'm checking with `glfwGetError((const char**))` but no errors are reported.
I build a band-aid for this bug to detect when the frames in come way too often. Then I'm switching to usleep: https://github.com/beaumanvienna/gfxRenderEngine/blob/0f5e6126e02764c9471eeec97e7562be1b911ef5/engine/platform/GLFW/GLGraphicsContext.cpp#L67 If there's anything else to check or any other feedback you need, please let me know. VSync works with an SDL window, so this is not a driver bug.

**Comment by HugoPeters1024** [comment]
Welcome to the party :P. I've since moved my interest to working with vulkan (where you have to do swapping yourself). The reason I thought this was a driver bug back in the day was because it only occurred on nvidia. But perhaps GLFW has some vendor specific code. Using usleep is a very reasonable solution imo.

**Comment by beaumanvienna** [comment]
I can reproduce this bug with an AMD/Radeon 570 card on Manjaro. 

Anyways, if you like to, you could contribute a Vulkan window to my rendering engine. That would be really cool. I understood drawing a simple triangle takes about 1500 lines of code in Vulkan. 

This is how the engine looks like so far: https://streamable.com/xy23ov 
It has batch rendering, sprite sheet management, a layering system, and game controller support.

**Comment by space8ar** [comment]
Hi 
I have Windows 7, GLFW version 3.3.5 and GPU is GT 630. I guess the problem lay in driver settings, like if you force some 3D settings by default via driver. In Windows I set "let the 3D application decide" in "3D Settings -> Adjust image settings with preview" and it helps in my case. So you may check your driver settings ("NVidia X Server Settings -> OpenGL Settings") and try different options
Good luck

**Comment by beaumanvienna** [comment]
Here's another suggestion for how to deal with this: [void GLContext::SwapBuffers()](https://github.com/beaumanvienna/gfxRenderEngine/blob/2074cd10260b86e41e04b18e1b96d05bea291bd5/engine/platform/OpenGL/GLGraphicsContext.cpp#L79) If the frame time is too short, this algo is switching to sleep().

**Comment by 9291Sam** [comment]
I was able to reproduce a very similar bug, where glfwSwapInterval(0) doesn't disable vSync, when glfwSwapInterval() was called in the scope of function B (see image). 
![image](https://user-images.githubusercontent.com/25729215/174504257-2f567a77-787c-4c6f-adde-bab54a21b550.png)

If I place glfwSwapInterval(0) into B, the function that lumps many render initialization functions together, it does not disable vSync. However, if I place it outside of the render function, anywhere in main(), it works. Quite a strange bug. If anyone wants code samples, please ask.

**Comment by 9291Sam** [comment]
Ok, I am writing this here incase anyone finds this in the future. It was a combo of a few different bugs, I'll try to explain. Thank you @elmindreda for pointing me in the correct direction!

At the point the previous picture was taken I was trying to move the initialize functions into another file, to make sure I hadn't broken anything too badly. I believe that because the function `render::initWindow()` was calling `glfwMakeContextCurrent(window)`, the context was being initialized in the scope of the function and then being destroyed as it went out of scope, thus creating a white window with its pointer pointing to uninitialized memory. (Yet more reasons to use smart pointers). Or I could be entirely wrong, I don't know I don't have this bit of code anymore. After about 30 minutes of trying to troubleshoot this, I just decided to cut my losses and start over and just wrap it all in a class (Which is probably what I *should* have done in the first place, but whatever, you get a little better every day).

**Comment by HugoPeters1024** [comment]
> I believe that because the function `render::initWindow()` was calling `glfwMakeContextCurrent(window)`, the context was being initialized in the scope of the function and then being destroyed as it went out of scope, thus creating a white window with its pointer pointing to uninitialized memory

That seems unlikely to me. makeContextCurrent does not add any values to stack and the window is a pointer (thus no deconstructors are called when they leave the scope)

**Comment by 9291Sam** [comment]
Yup, you're right. I did a bit more testing and I have thoroughly no idea what was going on. Thanks for the correction.

**Comment by zmn28hgbn59kcmlpio8unfh7fdre523esd28q9a** [comment]
Cannot reproduce, perhaps it was fixed long time ago.
